### PR TITLE
fix(runs): harden durable run contracts

### PR DESCRIPTION
**Pull request by kojiwakayama**
## Why

A module-level review of all 28 existing `src/runs` test steps found several false-confidence gaps:

- the injected project-env path rejected `__proto__`, `constructor`, and `prototype`, but the ambient-env path copied all three into task context;
- run events accepted fractional and negative IDs;
- run and schedule response schemas accepted negative durations and retries, plus zero or negative timeouts;
- an explicit API base URL ending in `/` generated `//runs` routes;
- the knowledge namespace test exercised only one of its three public methods;
- one eager test coupled run detail, event listing, and cancellation through response queue order;
- the public `veryfront/runs` runtime barrel had no contract test.

Direct probes reproduced the unsafe keys and invalid numeric values. Focused tests were added before each production fix and failed on the old behavior.

## What

- Apply the same unsafe-key policy to ambient and injected task environment sources.
- Require non-negative run durations and retry limits, positive run and source-schedule timeouts, and non-negative integer event IDs.
- Normalize trailing slashes before joining runs API paths.
- Execute all upload ID, upload path, and upload prefix knowledge-ingest variants with distinct expected request bodies.
- Split detail, event, and cancellation coverage into independent tests with exact URLs and methods.
- Replace manual host-env cleanup with the scoped test environment helper.
- Add a runtime export and owner-wiring contract for `veryfront/runs`.
- Reuse one authenticated client fixture for repeated request tests.
- Regenerate the runs API reference.

The existing `process` and `network` semantic disposition for `runs-client.test.ts` remains intentionally unchanged. The repository-sanctioned mock-fetch boundary still installs the outbound transport process-wide; replacing that seam belongs to the existing agent-runtime migration rather than a runs-only abstraction.

## Verification

- `deno task test:file src/runs`: 4 files, 37 steps passed
- `src/task/runner.test.ts`: 15 steps passed
- `cli/commands/schedule/handler.test.ts`: 32 steps passed
- `tests/docs/guide-examples.test.ts`: 47 steps passed
- `deno task test:unit:parallel`: 4,175 passed, 32,096 steps, 1 ignored
- `deno task test:unit:cwd`: 11 passed, 213 steps
- `deno task test:unit:cwd-exclusion`: 2 passed
- `deno task test:integration`: 315 passed, 2,923 steps
- focused `deno check`: passed
- `deno task typecheck`: passed
- `deno task lint`, `deno task fmt:check`: passed
- `deno task test:layout`: 2,357 executable tests, passed
- `deno task lint:test-semantic-dispositions`: 2,142 unit executables, 762 disposed files, passed
- test typecheck, anti-slop, sanitizer, cwd-read, module, dependency, API-doc, and public-doc gates: passed

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved API request handling when configured endpoints include trailing slashes.
  * Prevented unsafe, reserved, or invalid environment values from being passed into task contexts.
  * Strengthened validation for run durations, retries, timeouts, event IDs, and schedule settings.

* **Documentation**
  * Updated the Runs API reference with clearer schema descriptions, validation rules, and source links.

* **Tests**
  * Expanded coverage for API routes, configuration defaults, exports, environment filtering, and schema validation.

### CODE REVIEW

**Comment by kojiwakayama** [comment]
@codex review exact head d87aa7011. Focus on schema compatibility, unsafe ambient environment filtering, API base URL normalization, public barrel coverage, and whether the expanded client tests verify behavior without overfitting request internals.

**Review by kwakayama — APPROVED** [review]

**Review by kwakayama — DISMISSED** [review]
Independent Codex review of `d87aa7011ba0064135db028e66317eebe12afb1b` found no actionable code, spec, security, or regression issue.

The patch applies one unsafe-key policy to ambient and injected task env, tightens durable numeric response contracts, removes trailing API-base slashes without changing the authorized origin, and expands independent client/barrel coverage. Request bodies preserve the API's snake_case contract and remain JSON-encoded through the canonical transport. The mock-fetch seam is restored after each test and matches the repository-sanctioned process-wide boundary.

Verification under repository-pinned Deno 2.7.7:

- `deno task test:file src/runs`: 4 passed, 37 steps, 0 failed.
- `deno task test:file src/task/runner.test.ts cli/commands/schedule/handler.test.ts`: 5 passed, 47 steps, 0 failed.
- `deno check` on all 8 changed TypeScript files: passed.
- Targeted `deno lint` on all 8 changed TypeScript files: passed.
- Targeted `deno fmt --check` on 8 TypeScript files plus the generated runs reference: passed.
- Numeric probes confirmed rejection of `NaN`, infinities, and integers above `Number.MAX_SAFE_INTEGER` for every tightened field.
- `git diff --check` and changed-line security-pattern scan: passed.

Minor non-blocking deductions: the generated reference does not enumerate the new numeric refinements, and the ambient-env loop repeats part of the sanitizer decision instead of routing through one helper. Neither creates a current contract or security defect.

Score: 96/100. Verdict: APPROVE.

**Comment by kwakayama** [comment]
## Codex independent scored review

**Reviewed SHA:** `d87aa7011ba0064135db028e66317eebe12afb1b`

No actionable findings. The full pinned diff and surrounding runtime/client/schema contracts were reviewed for unsafe env keys and prototype pollution, numeric finiteness/integer/bounds, URL joining and origin authorization, request-body encoding, mock-fetch isolation, public barrel ownership, compatibility, and docs.

### Score

- Correctness and completeness: 40/40
- Regression tests and verification: 20/20
- Reliability and security: 15/15
- Repository standards and maintainability: 13/15
- Scope, documentation, and rollout clarity: 8/10

**Score: 96/100**

### Verification

- Pinned Deno 2.7.7 runs suite: 4 passed / 37 steps / 0 failed.
- Pinned related task and schedule suites: 5 passed / 47 steps / 0 failed.
- Pinned `deno check`: 8 changed TypeScript files passed.
- Pinned targeted lint and format checks: passed.
- Direct numeric adversarial probe: `NaN`, infinities, and unsafe integers rejected across all tightened fields.
- Exact-head CI snapshot: 40 passing, 0 pending, 0 failing, 6 expected skips.
- Review threads: 0 unresolved; pagination complete.

No LSP diagnostics service was available in the review environment; repository-pinned `deno check` was used as the native TypeScript diagnostic equivalent. The full suite was not rerun locally because exact-head CI already passed it.

Review-Gate:
Reviewer: Codex
Reviewed-SHA: d87aa7011ba0064135db028e66317eebe12afb1b
Score: 96/100
Actionable-Findings: 0
Verdict: APPROVE

**Comment by kwakayama** [comment]
Standards follow-up is prepared locally at `66de96f27376e8ea15ebd97af31334f299d8746b`, but GitHub rejected the guarded fast-forward because this PR is currently in the merge queue. Updating `test/audit-runs-module` requires dequeuing the PR first; no remote branch change was made.

Prepared fixes:
- document the durable run/event numeric constraints and regenerate `docs/api-reference/veryfront/runs.md` under Deno 2.7.7;
- route ambient task env through the shared workflow env sanitizer before applying the allowlist;
- add regression coverage proving ambient and existing workflow env share the unsafe/reserved/non-string policy.

Verification under pinned Deno 2.7.7:
- `deno task test:file src/runs`: 4 passed / 37 steps / 0 failed;
- full format and lint checks: passed;
- `deno check src/runs/index.ts`: passed;
- API reference freshness and validation: passed;
- `git diff --check`: passed.

The full pre-push unit suite reached 4,135 passed and failed only two unrelated timing-sensitive tests. Both failing files passed immediately in isolated reruns (2 files / 261 steps / 0 failed). Exact push blocker: `GH006: A pull request for this branch has been added to a merge queue. Branches that are queued for merging cannot be updated.`

**Review by kwakayama — APPROVED** [review]
> *This was generated by AI during triage.*\n\nReviewed the current diff, relevant base code, tests, and open threads. No actionable correctness, regression, security, compatibility, or material test-quality findings remain.

**Review by kwakayama — CHANGES_REQUESTED** [review]
Independent Codex exact-head review of `66de96f27376e8ea15ebd97af31334f299d8746b`.

[MEDIUM] `src/runs/runs-client.test.ts:691` — The negative missing-project test is host-environment dependent after suite-wide env cleanup was removed. If `VERYFRONT_PROJECT_SLUG` exists in the runner environment, `resolveProjectReference()` accepts it and the test proceeds to an outbound request instead of raising the asserted error. Under pinned Deno 2.7.7, `VERYFRONT_PROJECT_SLUG=ambient-project deno task test:file src/runs/runs-client.test.ts` reached `https://93.184.216.34/projects/ambient-project/runs` and timed out twice before the diagnostic run was stopped. Fix by explicitly scoping/unsetting `VERYFRONT_PROJECT_SLUG` for this negative test (or restoring deterministic suite isolation) and add a regression case that starts with an ambient slug but still exercises the missing-project branch.

No additional findings: schema sign/integer/finiteness/safe-integer bounds are correct; the RunSchema and RunEventSchema JSDoc-generated prose is current; ScheduleReferenceListSchema source JSDoc documents the positive timeout contract; ambient/injected/workflow env now reuse one sanitizer with regression coverage; and all original runs-client behavior remains covered while the eager route test is split and knowledge variants are expanded.

Verification: pinned Deno 2.7.7 runs suite passed 4 files / 37 steps; API-reference freshness and validation passed; deno check passed all 7 changed TypeScript files; targeted lint and format passed; direct numeric probes passed. CI remains in progress, so final green status is not yet available.

Score before severity cap: 93/100. Score after unresolved-MEDIUM cap: 84/100. Verdict: REQUEST_CHANGES.

**Comment by kwakayama** [comment]
Review-Gate:
Reviewer: Codex
Reviewed-SHA: 66de96f27376e8ea15ebd97af31334f299d8746b
Score: 84/100
Actionable-Findings: 1
Verdict: REQUEST_CHANGES

File: src/runs/runs-client.test.ts [hunk 1]
```diff
@@ -130,22 +129,23 @@ describe("VeryfrontRunsClient", () => {
 
   afterEach(() => {
     restoreMockFetch();
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/runs/runs-client.test.ts:131]: Fixed in `0a5f165711a3ddad514679c5f8aa22f63a62fe32`: the negative listing test scopes `VERYFRONT_PROJECT_SLUG` to empty with `withEnv`, installs an empty mock fetch, and asserts zero fetch calls. Pinned Deno 2.7.7 runs-client suite passes 19/19 steps; fmt, lint, check, and diff hygiene pass.

**Comment by kwakayama** [comment]
Addressed the remaining exact-head review finding in `0a5f165711a3ddad514679c5f8aa22f63a62fe32`.

- The missing-project negative test now scopes `VERYFRONT_PROJECT_SLUG` to empty with `withEnv`.
- It installs an empty mock fetch and asserts no request is made, so ambient host state cannot turn the local validation into network work.
- Pinned Deno 2.7.7 `runs-client.test.ts`: 19/19 steps passed.
- Changed-file format, lint, check, and `git diff --check` passed.
- Original thread replied to and resolved; pagination will be rechecked by the fresh exact-head review.

The prior standards fixes remain: shared env sanitization plus documented/generated numeric constraints.

**Review by kwakayama — APPROVED** [review]
Independent Codex exact-head review of 0a5f165711a3ddad514679c5f8aa22f63a62fe32.\n\nNo actionable findings. The full base-to-head diff preserves the original runs-client request and response contracts while tightening only invalid durable values. The latest negative-test fix deterministically scopes VERYFRONT_PROJECT_SLUG to empty, installs an empty fetch mock, and asserts zero calls, so ambient host state cannot bypass the missing-project validation. The shared workflow-env sanitizer now applies the unsafe/reserved/non-string policy to ambient, injected, and existing workflow values, with prototype-sensitive keys covered. Run, event, and source-schedule numeric constraints are reflected in generated API reference prose and targeted schema tests.\n\nReviewed areas: schema compatibility and numeric bounds, URL normalization and origin authorization, sanitizer behavior, public barrel wiring, knowledge-ingest variants, independent route tests, host-env isolation, security patterns, and repository standards. Review-thread pagination is complete: 1 total, 0 unresolved.\n\nScore:\n- Correctness and completeness: 40/40\n- Regression tests and verification: 18/20\n- Reliability and security: 15/15\n- Repository standards and maintainability: 15/15\n- Scope, documentation, and rollout clarity: 10/10\n\nScore: 98/100. Verdict: APPROVE.\n\nVerification gap: exact-head CI is still running (no failing checks observed at posting time). The repository is pinned to Deno 2.7.7; the available local binary is 2.9.4, so I did not treat its generic wrapper failure as pinned evidence. The PR's exact-head fix report records the focused pinned 2.7.7 runs-client suite at 19/19, and hosted Codex independently reported no issues on this SHA.

**Comment by kwakayama** [comment]
Review-Gate:\nReviewer: Codex\nReviewed-SHA: 0a5f165711a3ddad514679c5f8aa22f63a62fe32\nScore: 98/100\nActionable-Findings: 0\nVerdict: APPROVE
